Improve model generator error handling for non-existent models

AlexisChavez97 opened the pull request:
<!--
Thanks for contributing to Rails!

Please do not make *Draft* pull requests, as they still send
notifications to everyone watching the Rails repo.

Create a pull request when it is ready for review and feedback
from the Rails team :).

If your pull request affects documentation or any non-code
changes, guidelines for those changes are [available
here](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#contributing-to-the-rails-documentation)

About this template

The following template aims to help contributors write a good description for their pull requests.
We'd like you to provide a description of the changes in your pull request (i.e. bugs fixed or features added), the motivation behind the changes, and complete the checklist below before opening a pull request.

Feel free to discard it if you need to (e.g. when you just fix a typo). -->

### Motivation / Background
<!--
Describe why this Pull Request needs to be merged. What bug have you fixed? What feature have you added? Why is it important?
If you are fixing a specific issue, include "Fixes #ISSUE" (replace with the issue number, remove the quotes) and the issue will be linked to this PR.
-->

If you try to destroy a model that does not exist using the generator, you'll see a message like this one:
![image](https://github.com/user-attachments/assets/b3d57ed3-6809-40cf-85f1-59be17688bc5)

In the case of a typo, a developer could think the model was destroyed even though it wasn't.

This is the suggested behavior with these changes:
![image](https://github.com/user-attachments/assets/a9550c0f-9f97-458e-8493-ce04fe3119ef)

### Detail

This Pull Request adds a check to the `model_generator` to see if the model exists before deleting it

### Additional information

<!-- Provide additional information such as benchmarks, references to other repositories, or alternative solutions. -->

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [-] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

AlexisChavez97 committed:
Improve model generator error handling for non-existent models

$ rails d model NonExistentModel
Error: Model 'NonExistentModel' does not exist. Please choose an existing model or use --force to skip this check and run this generator again.

This change enhances the user experience when attempting to destroy
a non-existent model, providing a descriptive error message

Validate by class name instead of file path

Add helper to simulate constant load in test env

Integration test for model generator :revoke behavior

Fix generator integration tests for :revoke behavior

rasheedmhd commented:
Cool to see this here, I had saw it too and thought "Ha! this is weird", awesome  to know it has been worked on.

rasheedmhd commented:
Also, obviously this doesn't only occur when a typo is made but when the same command it ran more than once, the first time when the model still exist, the model gets deleted and also the subsequent run gives you the same message that the model has been deleted meanwhile it was already deleted and nothing happened for the second run.

johnreitano commented on railties/lib/rails/generators/named_base.rb:
```diff
@@ -201,6 +201,10 @@ def mountable_engine? # :doc:
           defined?(ENGINE_ROOT) && namespaced?
         end
 
+        def model_exists?
```
Looks like this method is not used

zzak commented:
So I think what we want is something like the inverse of `class_collisions`:
https://github.com/rails/rails/blob/e71a6d410e57589fc9e4b13c918eed427c936152/railties/lib/rails/generators/base.rb#L262-L284

Where we raise if the (model in this case) doesn't exist.

This could be useful in other scenarios, perhaps, and I think it's safer than looking at file paths.

AlexisChavez97 commented:
Thanks for taking a look @zzak, I updated the `model_exists` method to be the opposite of `class_collisions` and ran into a few issues in the generator tests. It seems like the `run_generator` helper doesn't load constants for the resources it generates and `model_exists` doesn't work correctly in the test environment unless the constants are loaded manually. I added a helper method to `run_generator` that forces the constants for the new resources to be loaded and that worked for the tests using the helper method. I'm not sure if this is something we want to do so let me know what you think

zzak commented:
@AlexisChavez97 Sorry if I sent you on a wild :goose: chase, I had assumed because `check_collision` has some heuristic to determine which constants are already defined, we could do the inverse and assume the class doesn't exist if there was no error.

Here's my attempt, but I think it demonstrates the same issue you ran into.

```diff
diff --git a/activerecord/lib/rails/generators/active_record/model/model_generator.rb b/activerecord/lib/rails/generators/active_record/model/model_generator.rb
index ae6adac6ad..e6cc55bc7e 100644
--- a/activerecord/lib/rails/generators/active_record/model/model_generator.rb
+++ b/activerecord/lib/rails/generators/active_record/model/model_generator.rb
@@ -8,6 +8,7 @@ class ModelGenerator < Base # :nodoc:
       argument :attributes, type: :array, default: [], banner: "field[:type][:index] field[:type][:index]"
 
       check_class_collision
+      check_class_exists
 
       class_option :migration, type: :boolean
       class_option :timestamps, type: :boolean
diff --git a/railties/lib/rails/generators/base.rb b/railties/lib/rails/generators/base.rb
index 63f5a1bb8a..431d0cfe98 100644
--- a/railties/lib/rails/generators/base.rb
+++ b/railties/lib/rails/generators/base.rb
@@ -283,6 +283,29 @@ def class_collisions(*class_names)
           end
         end
 
+        # Check whether the given class names exist already in the application
+        # or Ruby on Rails.
+        def class_exists(*class_names)
+          return unless behavior == :revoke
+          return if options.skip_collision_check?
+          return if options.force?
+
+          class_names.flatten.each do |class_name|
+            class_name = class_name.to_s
+            next if class_name.strip.empty?
+
+            # Split the class from its module nesting
+            nesting = class_name.split("::")
+            last_name = nesting.pop
+            last = extract_last_module(nesting)
+
+            unless last&.const_defined?(last_name.camelize, false)
+              raise Error, "The class '#{class_name}' does not exist. Please define it first or use --skip-collision-check "  \
+                           "or --force to skip this check and run this generator again."
+            end
+          end
+        end
+
         # Takes in an array of nested modules and extracts the last module
         def extract_last_module(nesting) # :doc:
           nesting.inject(Object) do |last_module, nest|
diff --git a/railties/lib/rails/generators/named_base.rb b/railties/lib/rails/generators/named_base.rb
index fe032f2552..afb542438f 100644
--- a/railties/lib/rails/generators/named_base.rb
+++ b/railties/lib/rails/generators/named_base.rb
@@ -222,6 +222,18 @@ def self.check_class_collision(options = {}) # :doc:
             class_collisions "#{options[:prefix]}#{name}#{options[:suffix]}"
           end
         end
+
+        def self.check_class_exists(options = {}) # :doc:
+          define_method :check_class_exists do
+            name = if respond_to?(:controller_class_name, true) # for ResourceHelpers
+              controller_class_name
+            else
+              class_name
+            end
+
+            class_exists "#{options[:prefix]}#{name}#{options[:suffix]}"
+          end
+        end
     end
   end
 end
diff --git a/railties/test/generators/model_generator_test.rb b/railties/test/generators/model_generator_test.rb
index bef8f9925a..e1eaf50fcc 100644
--- a/railties/test/generators/model_generator_test.rb
+++ b/railties/test/generators/model_generator_test.rb
@@ -354,11 +354,16 @@ def test_migration_error_is_not_shown_on_revoke
   end
 
   def test_migration_is_removed_on_revoke
-    run_generator
+    run_generator ["Account"]
     run_generator ["Account"], behavior: :revoke
     assert_no_migration "db/migrate/create_accounts.rb"
   end
 
+  def test_revoke_non_existent_model
+    error = capture(:stderr) { run_generator ["NonExistentModel"], behavior: :revoke }
+    assert_match(/Model 'NonExistentModel' does not exist/, error)
+  end
+
   def test_existing_migration_is_removed_on_force
     run_generator
     old_migration = Dir["#{destination_root}/db/migrate/*_create_accounts.rb"].first

```


```
bin/test test/generators/model_generator_test.rb
Run options: --seed 62406

# Running:

..F

Failure:
ModelGeneratorTest#test_revoke_non_existent_model [test/generators/model_generator_test.rb:364]:
Expected /Model 'NonExistentModel' does not exist/ to match "The class 'NonExistentModel' does not exist. Please define it first or use --skip-collision-check or --force to skip this check and run this generator again.\n".


bin/test test/generators/model_generator_test.rb:362

...The class 'Account' does not exist. Please define it first or use --skip-collision-check or --force to skip this check and run this generator again.
F

Failure:
ModelGeneratorTest#test_migration_is_removed_on_revoke [test/generators/model_generator_test.rb:359]:
Expected migration db/migrate/create_accounts.rb to not exist, but found /home/zzak/code/rails/railties/test/fixtures/tmp/db/migrate/20241031125539_create_accounts.rb.
Expected "/home/zzak/code/rails/railties/test/fixtures/tmp/db/migrate/20241031125539_create_accounts.rb" to be nil.


bin/test test/generators/model_generator_test.rb:356

............................................................

Finished in 1.914554s, 34.9951 runs/s, 196.3904 assertions/s.
67 runs, 376 assertions, 2 failures, 0 errors, 0 skips
```

AlexisChavez97 commented:
@zzak Yeah, that's the same issue I ran into in the tests environment. However I did test it in a test app and it is working as expected:

![image](https://github.com/user-attachments/assets/9e8a7b45-4b82-4009-9cfc-71937aa4f111)

This appears to be specific to the test environment where the classes aren't being loaded when using the `generators_test_helper`.  I found that forcing the class to be loaded inside the test 'fixes' the issue and makes tests pass:
`const_set(class_name, Class.new)`

Not sure if we would want to modify the `generators_test_helper` to adopt this behavior.

zzak commented:
@AlexisChavez97 I figured it out, we can use the isolated test pattern and pull in the `rails` command for an integration style test. I've made a branch here with the changes if you want to pull it in and test it out.

https://github.com/rails/rails/compare/main...zzak:re-53024

We don't necessarily need `Assertions` or `Behavior` but they give us `assert_migration` which I thought is nice to have. (This means we need the `destination` setup block too.

And the `test_migration_is_removed_on_revoke` unit test will never work unless we do some hack like force constants to load.

We've only used this in the model generator, but I could see it being useful if other commands implement the revoke/destroy API.

AlexisChavez97 commented:
@zzak Thanks for the guidance! I pulled your changes and tested them out - works great.

I did notice that there are about 15 failing tests across different generator files (resource, scaffold, namespaced) due to the new `class_exists` check. It seems to be the same constant loading issue we encountered with the model generator.

Would you recommend rewriting those tests using the isolated test pattern and do an integration test instead?

zzak commented:
@AlexisChavez97 Thanks for your patience, please give me a little more time to debug the tests myself. :bow:

AlexisChavez97 commented:
@zzak of course, let me know if there's anything I can do

zzak commented:
Rebased and pushed df8e585 which should fix the other tests, couple quick notes:

* `rails "generate", "whatever"` does not give us the same result as simply `Dir.chdir(app_path) { system("bin/rails generate whatever") }`
* Some of the tests were false positives or not testing the thing correctly, but this helped discover those. For example nested routes not having the correct contents.

Let me see about getting some feedback on this approach now that we have a green build. :pray:

AlexisChavez97 commented:
Hey @zzak, have there been any updates on this? 👀

zzak commented:
Just waiting on feedback, would you mind [squashing](https://edgeguides.rubyonrails.org/contributing_to_ruby_on_rails.html#squashing-commits) while we wait?
